src/models/dirac_mixture_multi_publisher_model.py

```diff
@@ -0,0 +1,632 @@
+        self, impressions: List[int], max_frequency: int = 1
+    ) -> ReachPoint:
+        """Predicts reach as a function of impressions, without single pub reach agreement.
+
+        Args:
+            impressions:  A length <p> list where p = #pubs.  The impression vector for
+                which we want to predict.
+            max_frequency: int, specifies the number of frequencies for which reach
+                will be reported.
+
+        Returns:
+            A ReachPoint specifying the predicted reach for this number of impressions.
+        """
+        self._fit()
+        hypothetical_direction = np.array(
+            [
+                hypothetical / baseline
+                for hypothetical, baseline in zip(impressions, self.baseline_imps)
+            ]
+        )
+        predicted_relative_freq_hist = self.optimizer.predict(
+            hypothetical_direction=hypothetical_direction,
+            customized_max_freq=max_frequency,
+        )
+        relative_kplus_reaches_from_zero = np.cumsum(
+            predicted_relative_freq_hist[::-1]
+        )[::-1]
+        kplus_reaches = (
+            (self.common_universe_size * relative_kplus_reaches_from_zero[1:])
+            .round(0)
+            .astype("int32")
+        )
+        return ReachPoint(impressions, kplus_reaches)
+
+    @staticmethod
+    def backsolve_impression(
+        curve: Callable, target_reach: int, starting_impression: int
+    ) -> int:
+        """From a reach curve, backsolves the impression from a target reach.
+
+        The backsolving is done by a customized bisection search.
+        This method is to be used in the later
+        `by_impressions_with_single_pub_reach_agreement` method.
+
+        Args:
+            curve:  A reach curve, i.e., function from impression (short for #impressions)
+                to reach.  Here reach only means the 1+ reach.
+            target_reach:  The target reach.
+            starting_impression:  Starting point when searching the impression.
+
+        Returns:
+            A value of impression such that curve(impression) equals to
+            target_reach.
+        """
+        if target_reach == 0:
+            return 0
+        probe = max(2, starting_impression)
+        count = 0
+        while curve(probe) < target_reach:
+            probe *= 2
+            count += 1
+            if count >= 10:
+                raise ValueError(
+                    "Cannot achieve target reach.  Does the given reach curve has a upper bound?"
+                )
+        right = probe
+        left = 0 if count == 0 else int(probe / 2)
+        # Loop termination criterion:  right - left <= 1, or either left or right hits
+        # the target reach.
+        while not (
+            curve(left) == target_reach
+            or curve(right) == target_reach
+            or right - left <= 1
+        ):
+            mid = (left + right) / 2
+            mid = int(round(mid))
```

Comment:
lines 531-532 -> mid = left + (right - left) // 2

Reply:
This is definitely cleaner.  Let me still keep the original codes since (1) it's correct although long (2) if left or right somehow become float, I'm a little afraid of potential bug (without any evidence!)